src/garage/torch/algos/dqn.py

```diff
@@ -0,0 +1,268 @@
+            qf_optimizer=torch.optim.Adam,
+            *,  # Everything after this is numbers.
+            steps_per_epoch=20,
+            n_train_steps=50,
+            max_episode_length_eval=None,
+            buffer_batch_size=64,
+            min_buffer_size=int(1e4),
+            discount=0.99,
+            qf_lr=_Default(1e-3),
+            clip_rewards=None,
+            clip_gradient=10,
+            target_update_freq=5,
+            reward_scale=1.):
+        self._clip_reward = clip_rewards
+        self._clip_grad = clip_gradient
+
+        self._steps_per_epoch = steps_per_epoch
+        self._target_update_freq = target_update_freq
+        self._episode_qf_losses = []
+        self._epoch_ys = []
+        self._epoch_qs = []
+
+        self._policy = policy
+        self._qf = qf
+        self._n_train_steps = n_train_steps
+
+        self._min_buffer_size = min_buffer_size
+        self._qf = qf
+        self._steps_per_epoch = steps_per_epoch
+        self._n_train_steps = n_train_steps
+        self._buffer_batch_size = buffer_batch_size
+        self._discount = discount
+        self._reward_scale = reward_scale
+        self.max_episode_length = env_spec.max_episode_length
+        self._max_episode_length_eval = (max_episode_length_eval
+                                         or self.max_episode_length)
+        self._episode_reward_mean = collections.deque(maxlen=100)
+
+        self.env_spec = env_spec
+        self.replay_buffer = replay_buffer
+        self.policy = policy
+        self.exploration_policy = exploration_policy
+
+        self._target_qf = copy.deepcopy(self._qf)
+        self._qf_optimizer = make_optimizer(qf_optimizer,
+                                            module=self._qf,
+                                            lr=qf_lr)
+        self._eval_env = eval_env
+        self.worker_cls = FragmentWorker
+
+    def train(self, trainer):
+        """Obtain samplers and start actual training for each epoch.
+
+        Args:
+            trainer (Trainer): Experiment trainer.
+
+        Returns:
+            float: The average return in last epoch cycle.
+
+        """
+        if not self._eval_env:
+            self._eval_env = trainer.get_env_copy()
+        last_returns = [float('nan')]
+        trainer.enable_logging = False
+
+        for _ in trainer.step_epochs():
+            for cycle in range(self._steps_per_epoch):
+                trainer.step_path = trainer.obtain_episodes(trainer.step_itr)
+                self.train_once(trainer.step_itr, trainer.step_path)
+                if (cycle == 0 and self.replay_buffer.n_transitions_stored >=
+                        self._min_buffer_size):
+                    trainer.enable_logging = True
+                    logger.log('Evaluating policy')
+
+                    epsilon_before = self.exploration_policy.epsilon
+                    eval_eps = obtain_evaluation_episodes(
+                        self.exploration_policy,
```

Comment:
Curious why use exploration_policy for evaluating. It seems that other algorithms use policy here.

Reply:
Depends on the algorithm and policy. All the DQN implementations evaluate with exploration noise because it actually helps the agent. There are certain edge cases where a deterministic agent can get stuck in an infinite loop if noise isn't added. See the beginning of this clip https://drive.google.com/file/d/1c3BQU5t9gUhWnS-ymzgJFP8R9tZCR_X0/view?usp=sharing

Reply:
Thanks for clarity. Should we also make this change to tf/DQN?

Reply:
Probably yes, but there's a bug in tf DQN right now. I'll add this change to that bug fix PR once its up.